diff --git a/hw/arm/xnuboot_sptm.c b/hw/arm/xnuboot_sptm.c
--- a/hw/arm/xnuboot_sptm.c
+++ b/hw/arm/xnuboot_sptm.c
@@ -31,6 +31,6 @@
 #define SPTM_EXPECTED_STRIDE    0x10000000
 
-#define GET_L2_PT_INDEX(a)      (( ((a)) & ( (BIT(36)-1)) ))
-#define STRIP_L2_PT_INDEX(a)    (( ((a)) & (~(BIT(36)-1)) ))
+#define GET_L2_PT_INDEX(a)      (( ((a)) & ( (BIT_ULL(36)-1)) ))
+#define STRIP_L2_PT_INDEX(a)    (( ((a)) & (~(BIT_ULL(36)-1)) ))
 
 static void alloc_ram(Object *cpuobj, hwaddr base, size_t len) {
