diff --git a/hw/arm/darwin.c b/hw/arm/darwin.c
--- a/hw/arm/darwin.c
+++ b/hw/arm/darwin.c
@@ -21,11 +21,9 @@
 #include "hw/arm/apple_amcc.h"
 #include "xnu/patch.h"
 #include "xnu/macho.h"
+#include "hw/intc/apple_aic.h"
 
 // See device tree specification section 2.3.8: ranges
 #define IO_RANGE_BASE_OFFSET     1
 
-// Use this for AICs where we don't know how many interrupts there should be
-#define A_GOOD_NUMBER_OF_IRQS    0x1000
-
 #define EXPECTED_FIRMWARE_NAME   "qemu-sptm"
@@ -42,4 +40,5 @@ struct DarwinState {
     MachineState parent_obj;
     ARMCPU *cpu;
+    DeviceState *aic;
     struct xnu_boot_info bootinfo;
 };
@@ -141,38 +140,4 @@ static void init_sep(struct dtree_node *dt_root) {
     alloc_zeroed("sep", sep_reg[0].base, sep_reg[0].len);
 }
 
-static void init_aic(struct dtree_node *dt_root, uint64_t iobase) {
-    // This is a bare-bones aic implementation that ignores all register reads/
-    // writes, and simply reports the correct number of IRQs
-    struct dtree_node *aic = adt_find_node(dt_root, "arm-io/aic");
-    struct adt_io_reg *aic_reg = adt_get_prop_val(aic, "reg");
-
-    uint64_t base = aic_reg[0].base + iobase;
-    alloc_zeroed("aic", base, aic_reg[0].len);
-
-    int aic_vers = -1;
-    sscanf(adt_get_prop_val(aic, "compatible"), "aic,%d", &aic_vers);
-
-    uint32_t num_irqs = 0;
-    switch (aic_vers) {
-        case 1:
-            // aic,1 needs register +0x04 from first iobase to report the number of interrupts
-            // This is always 8x the size of the ipid-mask for this device
-            num_irqs = 8 * adt_get_prop_len(aic, "ipid-mask");
-            address_space_write(&address_space_memory, base + 0x4, MEMTXATTRS_UNSPECIFIED, &num_irqs, sizeof(num_irqs));
-            break;
-
-        case 2:
-        case 3:
-            // aic,2 and aic,3 have num irqs at +0xC
-            num_irqs = A_GOOD_NUMBER_OF_IRQS;
-            address_space_write(&address_space_memory, base + 0xC, MEMTXATTRS_UNSPECIFIED, &num_irqs, sizeof(num_irqs));
-            break;
-
-        default:
-            fprintf(stderr, "warning: unsupported AIC, this will probably not work\n");
-            break;
-    }
-}
-
 static void setup_mte(Object *cpuobj, MachineState *machine, struct xnu_boot_info *info) {
@@ -258,7 +223,8 @@ static void darwin_init(MachineState *ms) {
     uint64_t iobase = arm_io_ranges[IO_RANGE_BASE_OFFSET];
     assert(0 == arm_io_ranges[0]); // child bus phys addr must be zero
 
+    s->aic = apple_aic_create(adt_find_node(dt_root, "arm-io/aic"), iobase);
+
     init_uart(dt_root, iobase);
-    init_aic(dt_root, iobase);
     init_sep(dt_root);
     init_cpu_impl(dt_root);
@@ -272,6 +238,8 @@ static void darwin_init(MachineState *ms) {
     qdev_realize(DEVICE(cpuobj), NULL, &error_fatal);
 
     // On Apple Si, FIQ is hardwired to platform timer
+    // Hardwire IRQ to the AIC
     qdev_connect_gpio_out(cpudev, GTIMER_HYPVIRT, qdev_get_gpio_in(cpudev, ARM_CPU_FIQ));
+    qdev_connect_gpio_out(s->aic, 0, qdev_get_gpio_in(cpudev, ARM_CPU_IRQ));
 
     qemu_register_reset(do_darwin_reset, s);
